Fix #2629. Add new config option latex_keep_old_macro_names

jfbu opened the pull request:
The cause of the issue was a clash about `\titleref` macro being already
defined in memoir class context. To avoid similar problems, make the
text styling macros

  `\strong`, `\code`, `\bfcode`, `\email`, `\tablecontinued`, `\titleref`,
  `\menuselection`, `\accelerator`, `\crossref`, `\termref`, `\optional`.

also available with `\sphinx` prefix, with a conf.py option (default
False) to only define them with this prefix. Thus backwards
compatibility is maintained.

On this occasion, some internal non-public macros have been renamed with
prefix `\spx@`, and some internal macros having public names (because
they are written by latex.py into the body of the latex document) have
been renamed to have `\sphinx` prefix.

The command `find . -name '*.sty' -exec grep -l \\\\spx@ {} \;` has
been executed in TeXLive 2015 and 2016 installations to check no package
has macros starting with `\spx@`.onlist, etc... environments.

The macros in sphinx.sty starting with \py@, or \DU, or \PYG have not
been modified. Similarly `\release`, `\version`, `\releasename`, etc...
have not been renamed.

The sphinxShadowBox environment used by topic directive is made
customizable via the introduction of three latex lengths variables.

jfbu committed:
Fix #2629. Add new config option ``latex_keep_old_macro_names``

The cause of the issue was a clash about ``\titleref`` macro being already
defined in memoir class context. To avoid similar problems, this makes
the text styling macros

  ``\strong``, ``\code``, ``\bfcode``, ``\email``, ``\tablecontinued``,
  ``\titleref``, ``\menuselection``, ``\accelerator``, ``\crossref``,
  ``\termref``, ``\optional``,

also available with ``\sphinx`` prefix, with a conf.py boolean option to
let sphinx.sty only define ``\sphinx``-prefixed macros. As default value
is False, backwards compatibility is maintained.

On this occasion, some internal non-public macros have been renamed with
prefix ``\spx@``. The command

    find . -name '*.sty' -exec grep -l \\\\spx@ {} \;

has been executed in TeXLive 2015 and 2016 installations to check no
package defines macros starting with ``\spx@``.

Some internal macros having public names (because they are written by
latex.py into the body of the latex document) have been renamed to have
``\sphinx`` prefix. The macros in sphinx.sty starting with \py@, or \DU,
or \PYG have not been modified. Similarly ``\release``, ``\version``,
``\releasename``, etc... have not been renamed.

tk0miya commented on CHANGES:
```diff
@@ -1,10 +1,16 @@
 ----------
 
 * #2676: (latex) Error with verbatim text in captions since Sphinx 1.4.4
+* new config option ``latex_macros_onlywithprefix`` to help resolve macro name conflicts with latex packages or classes such as memoir (ref #2629)
```
This is surely a bug fix, but it is also new feature. So the entry should move or copy to _Feature added_ section to let users know about the new option.

jfbu replied:
I propose to add in  _Feature added_:

```
* new config option ``latex_keep_old_macro_names`` with default value True.
  If False, let macros (for text styling) be defined only with ``\sphinx``-prefixed names.
  (may solve issues of conflict with user added latex packages).
```

tk0miya commented on sphinx/config.py:
```diff
@@ -209,6 +209,7 @@ class Config(object):
                            None),
         latex_logo = (None, None, string_classes),
         latex_appendices = ([], None),
+        latex_macros_onlywithprefix = (False, None),
```
This represents the behavior correctly, but it is hard to understand what the option does from its name.
How about `latex_provides_old_macros` or `latex_use_old_macros`?

Anyway, please add a description of the option to `doc/config.rst`.

jfbu replied:
indeed I have had big troubles with finding a name. The commit log explains that it applies only currently to the text styling macros. Your suggested names (which I agree are better) are opposite boolean, default value True. Perhaps `latex_keep_oldmacronames` ?

tk0miya commented:
At first, I wondered which branch is better to merge this. I'm afraid to break compatibilities by replacing new macros.
But now, after reviewing, I believe this is able to keep backward compatibilities.

So I'm okay to merge this into stable.
Please merge this after fixing some comments.

jfbu commented:
The changes `Sphinx-->sphinx` apply to macros and environments which have been introduced to Sphinx those last few months only and are not really public macros, although I noticed you mentioned `SphinxShadowBox` in closing an issue. Now it will be `sphinxShadowBox` environment. I think people bothered by that should be very few if at all any.

The documents will be produced with `\sphinxcode`,  `\sphinxcrossref` etc..., but (if the config boolean has default value) with `\def\sphinxcode{\code}`  etc..., so any user customization of `\code`,  `\crossref` ..., will be obeyed with no break of backwards compatibility. (as LaTeX is not like a procedural language, it doesn't matter that number of parameters play no rôle in `\def\sphinxcode{\code}`, the parameters are fetched later by `\code` once it has replaced `\sphinxcode` in token stream).

tk0miya commented:
Looks great!

jfbu committed:
Make sphinxShadowBox customizable via three lengths with public names

jfbu commented:
I have squashed and re-organized into two successive commits. Will merge into stable now.

jfbu commented:
Memo: the commit log at 5510653 speaks about "default value is False", the context is clear that this means that by default the old macro names are still used. But the config boolean deciding this is the negative of the one originally referred to in the commit log, after some rebase and edit and split. Actual files have the correct wording.

jfbu commented:
@tk0miya after this, stable branch does not merge cleanly anymore into master. I have at branch at https://github.com/jfbu/sphinx/tree/jfbu_formerging4cf78e6 which solved conflicts to show how one could merge current stable into current master. I am not experimented enough in git to understand how one handles this (if I merge this branch into master it would be a fast forward, but then future attempt to merge stable will have the original conflicts).

jfbu commented:
perhaps I should add a commit to master reverting some of the changes since last stable merge, in order for git merge stable then to apply without conflicts ?

tk0miya commented:
The conflicts comes from the changes on both branches. I think all changes have its meaning for each commit. So reverting is not good way. We have to resolve all of conflicts by manually. 

Would you like me to merge them if you are not good at?

jfbu commented:
I can merge stable into master, resolving conflicts in passing. however I am afraid conflicts will come back next time future tip of stable is merged into master ? (I just read about [git rerere](https://git-scm.com/blog/2010/03/08/rerere.html) but I don't know practice of Sphinx in such cases)

tk0miya commented:
The conflicts will comes if we modified same points on both branches after this merging.
So usually, there would be no conflicts (or small ones).

jfbu commented:
ok, thanks for reassuring. I have now merged stable into master. And I then did an additional commit to change SphinxVerbatim environment name to sphinxVerbatim, in line with recent conventions for the other environments. In my first tries this morning I did that during merge, but I think it is better to separate the commit.
